addons/beehave/debug/graph_edit.gd

```diff
@@ -0,0 +1,219 @@
+@tool
+extends GraphEdit
+
+const BeehaveGraphNode := preload("graph_node.gd")
+
+const PORT_OFFSET := Vector2(-1, 0)
+const PROGRESS_SHIFT: int = 50
+const INACTIVE_COLOR: Color = Color("#898989aa")
+const ACTIVE_COLOR: Color = Color("#ffcc00c8")
+const SUCCESS_COLOR: Color = Color("#009944c8")
+
+const ICONS := {
+	"SelectorComposite": preload("../icons/selector.svg"),
+	"SelectorRandomComposite": preload("../icons/selector_random.svg"),
+	"SelectorStarComposite": preload("../icons/selector_reactive.svg"),
+	"ActionLeaf": preload("../icons/action.svg"),
+	"BeehaveTree": preload("../icons/tree.svg"),
+	"ConditionLeaf": preload("../icons/condition.svg"),
+	"LimiterDecorator": preload("../icons/limiter.svg"),
+	"InverterDecorator": preload("../icons/inverter.svg"),
+	"SequenceComposite": preload("../icons/sequence.svg"),
+	"SequenceRandomComposite": preload("../icons/sequence_random.svg"),
+	"SequenceStarComposite": preload("../icons/sequence_reactive.svg"),
+	"AlwaysSucceedDecorator": preload("../icons/succeeder.svg"),
+	"AlwaysFailDecorator": preload("../icons/failer.svg")
+}
+
+
+var updating_graph: bool = false
+var beehave_tree: Dictionary:
+	set(value):
+		if beehave_tree == value:
+			return
+		beehave_tree = value
+		_update_graph()
+
+var active_nodes: Array[String]
+var progress: int = 0
+
+
+func _ready() -> void:
+	custom_minimum_size = Vector2(100, 300)
+	arrange_nodes_button_hidden = true
+	minimap_enabled = false
+
+
+func _update_graph() -> void:
+	if updating_graph:
+		return
+
+	updating_graph = true
+
+	clear_connections()
+
+	for child in get_children():
+		remove_child(child)
+		child.queue_free()
+
+	if not beehave_tree.is_empty():
+		_add_nodes(beehave_tree)
+		_connect_nodes(beehave_tree)
+		_arrange_nodes(beehave_tree)
+
+	updating_graph = false
+
+
+func _add_nodes(node: Dictionary) -> void:
```

Comment:
It is not guaranteed that `node` has any data (especially when switching scenes). When I ran it while switching scenes it gave me an error that node` does not contain things like `name`. I solved the issue by doing this at the start of this method:
```suggestion
func _add_nodes(node: Dictionary) -> void:
   if node.is_empty():
	return
```